

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genetest.statistics.models.linear &mdash; genetest 0.4.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../../../genindex.html" />
    <link rel="search" title="Search" href="../../../../search.html" /> 

  
  <script src="../../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../../index.html" class="icon icon-home"> genetest
          

          
          </a>

          
            
            
              <div class="version">
                0.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../../module_content/modules.html">genetest</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../../index.html">genetest</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../../index.html">Module code</a> &raquo;</li>
        
          <li><a href="../../../genetest.html">genetest</a> &raquo;</li>
        
      <li>genetest.statistics.models.linear</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genetest.statistics.models.linear</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="c1"># This file is part of genetest.</span>
<span class="c1">#</span>
<span class="c1"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c1"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c1"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c1"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">statsmodels.api</span> <span class="k">as</span> <span class="nn">sm</span>

<span class="kn">from</span> <span class="nn">..core</span> <span class="k">import</span> <span class="n">StatsModels</span><span class="p">,</span> <span class="n">StatsError</span>


<span class="n">__copyright__</span> <span class="o">=</span> <span class="s2">&quot;Copyright 2016, Beaulieu-Saucier Pharmacogenomics Centre&quot;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s2">&quot;Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)&quot;</span>


<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;StatsLinear&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="StatsLinear"><a class="viewcode-back" href="../../../../module_content/genetest.statistics.models.html#genetest.statistics.models.linear.StatsLinear">[docs]</a><span class="k">class</span> <span class="nc">StatsLinear</span><span class="p">(</span><span class="n">StatsModels</span><span class="p">):</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">condition_value_t</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="n">eigenvals_t</span><span class="o">=</span><span class="mf">1e-10</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Initializes a &#39;StatsLinear&#39; instance.</span>

<span class="sd">        Args:</span>
<span class="sd">            condition_value_t (int): The condition value threshold (for</span>
<span class="sd">                                     multicollinearity). Usually, values higher</span>
<span class="sd">                                     than 1000 indicate strong</span>
<span class="sd">                                     multicollinearity or other numerical</span>
<span class="sd">                                     problems.</span>
<span class="sd">            eigenvals_t (float): The Eigen value threshold (for</span>
<span class="sd">                                 multicollinearity). Usually, values lower than</span>
<span class="sd">                                 1e-10 might indicate strong multicollinearity</span>
<span class="sd">                                 or singular design matrix.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Saving the condition value threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_condition_value_t</span> <span class="o">=</span> <span class="n">condition_value_t</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_eigenvals_t</span> <span class="o">=</span> <span class="n">eigenvals_t</span>

    <span class="k">def</span> <span class="nf">_results_handler</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fitted</span><span class="p">):</span>
        <span class="c1"># Checking the condition number (according to StatsModels, condition</span>
        <span class="c1"># number higher than 1000 indicate that there are strong</span>
        <span class="c1"># multicollinearity or other numerical problems)</span>
        <span class="k">if</span> <span class="n">fitted</span><span class="o">.</span><span class="n">condition_number</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_condition_value_t</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="s2">&quot;condition number is large, </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">fitted</span><span class="o">.</span><span class="n">condition_number</span><span class="p">,</span>
            <span class="p">))</span>

        <span class="c1"># Checking the smallest eigenvalue (according to StatsModels, values</span>
        <span class="c1"># lower than 1e-10 might indicate that there are strong</span>
        <span class="c1"># multicollinearity problems or that the design matrix is singular)</span>
        <span class="k">if</span> <span class="n">fitted</span><span class="o">.</span><span class="n">eigenvals</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_eigenvals_t</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="s2">&quot;smallest eigenvalue is small, </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">fitted</span><span class="o">.</span><span class="n">eigenvals</span><span class="o">.</span><span class="n">min</span><span class="p">(),</span>
            <span class="p">))</span>

        <span class="c1"># Results about the model fit.</span>
        <span class="n">out</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;MODEL&quot;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;r_squared_adj&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">rsquared_adj</span><span class="p">,</span>
                <span class="s2">&quot;log_likelihood&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">llf</span><span class="p">,</span>
                <span class="s2">&quot;nobs&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">nobs</span>
            <span class="p">}</span>
        <span class="p">}</span>

        <span class="c1"># Getting the confidence intervals</span>
        <span class="n">conf_ints</span> <span class="o">=</span> <span class="n">fitted</span><span class="o">.</span><span class="n">conf_int</span><span class="p">()</span>

        <span class="c1"># Results about individual model parameters.</span>
        <span class="n">parameters</span> <span class="o">=</span> <span class="n">fitted</span><span class="o">.</span><span class="n">params</span><span class="o">.</span><span class="n">index</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="n">parameters</span><span class="p">:</span>
            <span class="c1"># If GWAS, check that inference could be done on the SNP.</span>
            <span class="k">if</span> <span class="n">param</span> <span class="o">==</span> <span class="s2">&quot;SNPs&quot;</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">fitted</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="n">param</span><span class="p">]):</span>
                <span class="k">raise</span> <span class="n">StatsError</span><span class="p">(</span><span class="s2">&quot;Inference did not converge.&quot;</span><span class="p">)</span>

            <span class="n">out</span><span class="p">[</span><span class="n">param</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;coef&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">params</span><span class="p">[</span><span class="n">param</span><span class="p">],</span>
                <span class="s2">&quot;std_err&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">bse</span><span class="p">[</span><span class="n">param</span><span class="p">],</span>
                <span class="s2">&quot;lower_ci&quot;</span><span class="p">:</span> <span class="n">conf_ints</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">param</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
                <span class="s2">&quot;upper_ci&quot;</span><span class="p">:</span> <span class="n">conf_ints</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">param</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span>
                <span class="s2">&quot;t_value&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">tvalues</span><span class="p">[</span><span class="n">param</span><span class="p">],</span>
                <span class="s2">&quot;p_value&quot;</span><span class="p">:</span> <span class="n">fitted</span><span class="o">.</span><span class="n">pvalues</span><span class="p">[</span><span class="n">param</span><span class="p">],</span>
            <span class="p">}</span>

        <span class="k">return</span> <span class="n">out</span>

<div class="viewcode-block" id="StatsLinear.fit"><a class="viewcode-back" href="../../../../module_content/genetest.statistics.models.html#genetest.statistics.models.linear.StatsLinear.fit">[docs]</a>    <span class="k">def</span> <span class="nf">fit</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">X</span><span class="p">,</span> <span class="n">handler</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Fit the model.</span>

<span class="sd">        Args:</span>
<span class="sd">            y (pandas.DataFrame): The vector of endogenous variable.</span>
<span class="sd">            X (pandas.DataFrame): The matrix of exogenous variables.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Creating the OLS model from StatsModels and fitting it</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">sm</span><span class="o">.</span><span class="n">OLS</span><span class="p">(</span><span class="n">y</span><span class="p">,</span> <span class="n">X</span><span class="p">)</span>
        <span class="n">handler</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_results_handler</span> <span class="k">if</span> <span class="n">handler</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">handler</span>
        <span class="k">return</span> <span class="n">handler</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">())</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Beaulieu-Saucier Pharmacogenomics Centre.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../../',
            VERSION:'0.4.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="../../../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>