

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>genetest.statistics.descriptive &mdash; genetest 0.4.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> genetest
          

          
          </a>

          
            
            
              <div class="version">
                0.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../analysis.html">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../module_content/modules.html">genetest</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">genetest</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
          <li><a href="../../genetest.html">genetest</a> &raquo;</li>
        
      <li>genetest.statistics.descriptive</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for genetest.statistics.descriptive</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>


<span class="c1"># This file is part of genetest.</span>
<span class="c1">#</span>
<span class="c1"># This work is licensed under the Creative Commons Attribution-NonCommercial</span>
<span class="c1"># 4.0 International License. To view a copy of this license, visit</span>
<span class="c1"># http://creativecommons.org/licenses/by-nc/4.0/ or send a letter to Creative</span>
<span class="c1"># Commons, PO Box 1866, Mountain View, CA 94042, USA.</span>


<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="n">__all__</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;get_maf&quot;</span><span class="p">]</span>


<div class="viewcode-block" id="get_maf"><a class="viewcode-back" href="../../../module_content/genetest.statistics.html#genetest.statistics.descriptive.get_maf">[docs]</a><span class="k">def</span> <span class="nf">get_maf</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the minor allele frequency using genotypes.</span>

<span class="sd">    Args:</span>
<span class="sd">        genotypes (pandas.Series): The genotypes.</span>
<span class="sd">        minor (str): The minor allele.</span>
<span class="sd">        major (str): The major allele.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Returns the MAF, the minor allele, the major allele and a</span>
<span class="sd">               boolean telling if the markers were flip or not.</span>

<span class="sd">    Note</span>
<span class="sd">    ====</span>
<span class="sd">        The frequency is computed using the alternative allele (i.e. the ``2``</span>
<span class="sd">        genotype). If there are more ``2`` genotypes than ``0`` genotypes, the</span>
<span class="sd">        alternative allele frequency will be higher than 0.5.</span>

<span class="sd">    Note</span>
<span class="sd">    ====</span>
<span class="sd">        When computing the alternative allele frequency, the missing genotypes</span>
<span class="sd">        are excluded. If there are no genotypes, ``NaN`` is returned.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nanmean</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>

    <span class="k">if</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">maf</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="kc">False</span></div>


<span class="k">def</span> <span class="nf">get_sex_maf</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">sexes</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Computes the minor allele frequency for sexual chromosomes.</span>

<span class="sd">    Args:</span>
<span class="sd">        genotypes (pandas.Series): The genotypes.</span>
<span class="sd">        sexes (pandas.Series): The sex (0=female, 1=male).</span>
<span class="sd">        minor (str): The minor allele.</span>
<span class="sd">        major (str): The major allele.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: Returns the MAF, the minor allele, the major allele and a</span>
<span class="sd">               boolean telling if the markers were flip or not.</span>

<span class="sd">    Warning</span>
<span class="sd">    =======</span>
<span class="sd">        The ``sexes`` vector should have the same order as the genotypes one,</span>
<span class="sd">        and the males are coded as ``1``, while females are coded as ``0``.</span>
<span class="sd">        Missing sexes should be coded as ``NaN``.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Keeping only the non-missing data</span>
    <span class="n">not_missing</span> <span class="o">=</span> <span class="o">~</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span> <span class="o">|</span> <span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">sexes</span><span class="p">))</span>
    <span class="k">if</span> <span class="p">(</span><span class="o">~</span><span class="n">not_missing</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="kc">False</span>

    <span class="c1"># keeping only non missing values</span>
    <span class="n">genotypes</span> <span class="o">=</span> <span class="n">genotypes</span><span class="p">[</span><span class="n">not_missing</span><span class="p">]</span>
    <span class="n">sexes</span> <span class="o">=</span> <span class="n">sexes</span><span class="p">[</span><span class="n">not_missing</span><span class="p">]</span>

    <span class="c1"># The number of males and females</span>
    <span class="n">nb_males</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sexes</span><span class="p">)</span>
    <span class="n">nb_females</span> <span class="o">=</span> <span class="n">sexes</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">nb_males</span>

    <span class="c1"># Computing the MAF</span>
    <span class="n">maf</span> <span class="o">=</span> <span class="p">(</span>
        <span class="p">(</span><span class="o">-</span><span class="mf">0.5</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">genotypes</span><span class="p">,</span> <span class="n">sexes</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">genotypes</span><span class="p">))</span>
        <span class="o">/</span> <span class="p">(</span><span class="n">nb_males</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">nb_females</span><span class="p">))</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.5</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">maf</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="kc">True</span>

    <span class="k">return</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor</span><span class="p">,</span> <span class="n">major</span><span class="p">,</span> <span class="kc">False</span>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Beaulieu-Saucier Pharmacogenomics Centre.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'0.4.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="../../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>